File: incre-tests/synduce/constraints/ensures/mpss_noe.f
Inductive List = elt Int | cons {Int, List};
Inductive NList = line List | ncons {List, NList};
Inductive CNList = sglt List | cat {CNList, CNList};

cton = fix (
  \f: CNList -> NList.
  let dec = fix (
    \g: CNList -> CNList -> NList. \l: CNList. \c: CNList.
    match c with
      sglt x -> ncons {x, f l}
    | cat {x, y} -> g (cat {y, l}) x
    end
  ) in \c: CNList.
  match c with
    sglt x -> line x
  | cat {x, y} -> dec y x
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> + h (f t)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

spec = \xs: NList. (fix (
  \f: NList -> {Int, Int}. \xs: NList.
  match xs with
    line a -> {max 0 (sum a), sum a}
  | ncons {h, t} ->
    let hsum = sum h in
      let tres = f t in
        {max (+ tres.2 hsum) tres.1, + tres.2 hsum}
  end
) xs).1;

target = fix (
  \f: CNList -> Compress CNList. \c: CNList.
  match c with
    sglt x ->
    let info = sum x in /*This invocation is provided in Synduce's template*/
      c
  | cat {l, r} -> cat {f l, f r}
  end
);

main = \c: CNList. spec (cton (target c));

File: ml-benchmark/synduce/constraints/ensures/mpss_noe.ml
type list = Elt of int | Cons of int * list
type nlist = Line of list | Ncons of list * nlist
type cnlist = Sglt of list | Cat of cnlist * cnlist

let rec cton c =
  let rec dec l c =
    match c with
    | Sglt x -> Ncons (x, cton l)
    | Cat (x, y) -> dec (Cat (y, l)) x
  in
  match c with
  | Sglt x -> Line x
  | Cat (x, y) -> dec y x

let rec sum xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> h + sum t

let max a b = if a < b then b else a

let spec xs =
  let rec f xs =
    match xs with
    | Line a -> (max 0 (sum a), sum a)
    | Ncons (h, t) ->
      let hsum = sum h in
      let tres = f t in
      match tres with
      | (t1, t2) -> (max (t2 + hsum) t1, t2 + hsum)
  in
  let res = f xs in
  match res with
  | (r1, r2) -> r1

val target: cnlist -> cnlist compress
let rec target c =
  match c with
  | Sglt x ->
    let info = sum x in
    c
  | Cat (l, r) -> Cat (target l, target r)

let program c = spec (cton (target c))